src/org/traccar/protocol/GranitProtocolDecoder.java

```diff
@@ -0,0 +1,212 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.jboss.netty.channel.Channel;
+import org.traccar.BaseProtocolDecoder;
+import org.traccar.helper.BitUtil;
+import org.traccar.helper.Checksum;
+import org.traccar.model.Position;
+
+import java.net.SocketAddress;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+public class GranitProtocolDecoder extends BaseProtocolDecoder {
+
+    private static final int HEADER_LENGTH = 6;
+
+    public GranitProtocolDecoder(GranitProtocol protocol) {
+        super(protocol);
+    }
+
+    public static void appendChecksum(ChannelBuffer buffer, int length) {
+        buffer.writeByte('*');
+        int checksum = Checksum.xor(buffer.toByteBuffer(0, length)) & 0xFF;
+        String checksumString = String.format("%02X", checksum);
+        buffer.writeBytes(checksumString.getBytes(StandardCharsets.US_ASCII));
+        buffer.writeByte('\r'); buffer.writeByte('\n');
+    }
+
+    private static void sendResponseCurrent(Channel channel, int deviceId, long time) {
+        ChannelBuffer response = ChannelBuffers.dynamicBuffer(ByteOrder.LITTLE_ENDIAN, 0);
+        response.writeBytes("BB+UGRC~".getBytes(StandardCharsets.US_ASCII));
+        response.writeShort(6); //binary length
+        response.writeInt((int) time);
+        response.writeShort(deviceId);
+        appendChecksum(response, 16);
+        channel.write(response);
+    }
+
+    private static void sendResponseArchive(Channel channel, int deviceId, int packNum) {
+        ChannelBuffer response = ChannelBuffers.dynamicBuffer(ByteOrder.LITTLE_ENDIAN, 0);
+        response.writeBytes("BB+ARCF~".getBytes(StandardCharsets.US_ASCII));
+        response.writeShort(4); //binary length
+        response.writeShort(packNum);
+        response.writeShort(deviceId);
+        appendChecksum(response, 14);
+        channel.write(response);
+    }
+
+    private static void decodeStructure(ChannelBuffer buf, Position position) {
+        short flags = buf.readUnsignedByte();
+        position.setValid(BitUtil.check(flags, 7));
+        position.set(Position.KEY_ALARM, BitUtil.check(flags, 1));
+
+        short satDel = buf.readUnsignedByte();
+        position.set(Position.KEY_SATELLITES, BitUtil.from(satDel, 4));
+
+        int pdop = BitUtil.to(satDel, 4);
+        position.set("pdop", pdop);
+
+        int lonDegrees = buf.readUnsignedByte();
+        int latDegrees = buf.readUnsignedByte();
+        int lonMinutes = buf.readUnsignedShort();
+        int latMinutes = buf.readUnsignedShort();
+        double latitude = latDegrees + (double) latMinutes / 60000;
```

Comment:
`(double) latMinutes / 60000` can be replaced with `latMinutes / 60000.0` for readability